experimental/dds/tree2/src/shared-tree-core/branch.ts

```diff
@@ -315,49 +314,74 @@ export class SharedTreeBranch<TEditor extends ChangeFamilyEditor, TChange> exten
 	 */
-	public undo(): [change: TChange, newCommit: GraphCommit<TChange>] | undefined {
-		assert(
-			this.undoRedoManager !== undefined,
-			0x686 /* Must construct branch with an `UndoRedoManager` in order to undo. */,
-		);
-		// TODO: allow this once it becomes possible to compose the changesets created by edits made
-		// within transactions and edits that represent completed transactions.
-		assert(!this.isTransacting(), 0x66a /* Undo is not yet supported during transactions */);
-
-		const undoChange = this.undoRedoManager?.undo(this.getHead());
-		if (undoChange !== undefined) {
-			return this.applyChange(undoChange, mintRevisionTag(), LocalCommitSource.Undo);
+	public updateRevertibleCommit(commit: GraphCommit<TChange>) {
+		if (this.revertibleCommits.get(commit.revision) !== undefined) {
+			this.revertibleCommits.set(commit.revision, commit);
 		}
+	}
 
-		return undefined;
+	private makeSharedTreeRevertible(
+		commit: GraphCommit<TChange>,
+		kind: RevertibleKind,
+	): Revertible {
+		this.revertibleCommits.set(commit.revision, commit);
+		const revertible = {
+			kind,
+			origin: {
+				// This is currently always the case, but we may want to support reverting remote ops
+				isLocal: true,
+			},
+			revert: () => {
+				this.revert(commit.revision, kind);
+				return RevertResult.Success;
+			},
+			discard: () => {
+				// TODO: delete the repair data from the forest
+				this.revertibleCommits.delete(commit.revision);
+				this.revertibles.delete(revertible);
+				return DiscardResult.Success;
+			},
+		};
+		this.revertibles.add(revertible);
+		return revertible;
 	}
 
-	/**
-	 * Redoes the last change made by the client. If there is no change to redo then this method has no effect.
-	 * This method will error if no {@link UndoRedoManager} was provided when this branch was constructed.
-	 * It is invalid to call this method while a transaction is open (this will be supported in the future).
-	 * @returns the change to this branch and the new head commit, or undefined if there was nothing to redo
-	 */
-	public redo(): [change: TChange, newCommit: GraphCommit<TChange>] | undefined {
-		assert(
-			this.undoRedoManager !== undefined,
-			0x687 /* Must construct branch with an `UndoRedoManager` in order to redo. */,
-		);
-		// TODO: allow this once it becomes possible to compose the changesets created by edits made
-		// within transactions and edits that represent completed transactions.
-		assert(!this.isTransacting(), 0x67e /* Redo is not yet supported during transactions */);
+	private revert(
+		revision: RevisionTag,
+		revertibleKind: RevertibleKind,
+	): [change: TChange, newCommit: GraphCommit<TChange>] | undefined {
+		assert(!this.isTransacting(), "Undo is not yet supported during transactions");
+
+		const commit = this.revertibleCommits.get(revision);
+		assert(commit !== undefined, "expected to find a revertible commit");
```

Comment:
This can happen if you call discard and then revert on the same revertible. We should probably prevent this in the revertible itself by only allowing one call to discard or revert, and then every subsequent call fails.